File: src/com/intellij/idea/plugin/hybris/type/system/meta/impl/TSMetaEnumImpl.java
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2014-2016 Alexander Bartash <AlexanderBartash@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.type.system.meta.impl;

import com.intellij.idea.plugin.hybris.type.system.meta.TSMetaEnum;
import com.intellij.idea.plugin.hybris.type.system.meta.impl.CaseInsensitive.NoCaseMultiMap;
import com.intellij.idea.plugin.hybris.type.system.model.EnumType;
import com.intellij.idea.plugin.hybris.type.system.model.EnumValue;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.project.Project;
import com.intellij.util.xml.DomAnchor;
import com.intellij.util.xml.DomService;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Stream;

public class TSMetaEnumImpl extends TSMetaEntityImpl<EnumType> implements TSMetaEnum {

    private final NoCaseMultiMap<TSMetaEnumValue> name2ValueObj = new NoCaseMultiMap<>();
    private final Set<DomAnchor<EnumType>> myAllDoms = new LinkedHashSet<>();
    private final boolean myAutoCreate;
    private final boolean myGenerate;
    private final boolean myDynamic;
    private final String myDescription;
    private final String myJaloClass;

    public TSMetaEnumImpl(final Module module, final Project project, final String name, final EnumType dom, final boolean custom) {
        super(module, project, name, dom, custom);
        myAutoCreate = Boolean.TRUE.equals(dom.getAutoCreate().getValue());
        myGenerate = Boolean.TRUE.equals(dom.getGenerate().getValue());
        myDynamic = Boolean.TRUE.equals(dom.getDynamic().getValue());
        myDescription = dom.getDescription().getStringValue();
        myJaloClass = dom.getJaloClass().getStringValue();

        myAllDoms.add(DomService.getInstance().createAnchor(dom));
    }

    @NotNull
    @Override
    public Collection<? extends TSMetaEnumValue> getValues() {
        return name2ValueObj.values();
    }

    @NotNull
    @Override
    public Collection<? extends TSMetaEnumValue> findValueByName(@NotNull final String name) {
        return new ArrayList<>(name2ValueObj.get(name));
    }

    @Override
    public void createValue(final @NotNull EnumValue domEnumValue) {
        final TSMetaEnumValue result = new TSMetaEnumValueImpl(getModule(), getProject(), this, domEnumValue, isCustom());

        if (result.getName() != null) {
            name2ValueObj.putValue(result.getName(), result);
        }
    }

    @Override
    public boolean isAutoCreate() {
        return myAutoCreate;
    }

    @Override
    public boolean isGenerate() {
        return myGenerate;
    }

    @Override
    public boolean isDynamic() {
        return myDynamic;
    }

    @Override
    public String getDescription() {
        return myDescription;
    }

    @Override
    public String getJaloClass() {
        return myJaloClass;
    }

    @Override
    public void merge(final TSMetaEnum another) {
        another.getValues().stream()
               .filter(anotherValue -> anotherValue.getName() != null)
               .forEach(anotherValue -> name2ValueObj.putValue(anotherValue.getName(), anotherValue));

        addDomRepresentation(another.retrieveDom());
    }

    @NotNull
    @Override
    public Stream<? extends EnumType> retrieveAllDomsStream() {
        return myAllDoms.stream()
                        .map(DomAnchor::retrieveDomElement)
                        .filter(Objects::nonNull);
    }

    protected void addDomRepresentation(final @NotNull EnumType anotherDom) {
        myAllDoms.add(DomService.getInstance().createAnchor(anotherDom));
    }

    public static class TSMetaEnumValueImpl extends TSMetaEntityImpl<EnumValue> implements TSMetaEnumValue {

        private final TSMetaEnum myOwner;
        private final String myDescription;

        public TSMetaEnumValueImpl(final Module module, final Project project, final @NotNull TSMetaEnum owner, final @NotNull EnumValue dom, final boolean custom) {
            super(module, project, extractEnumValueName(dom), dom, custom);
            myOwner = owner;
            myDescription = dom.getDescription().getStringValue();
        }

        @Nullable
        private static String extractEnumValueName(@NotNull final EnumValue dom) {
            return dom.getCode().getStringValue();
        }

        @Override
        public String getDescription() {
            return myDescription;
        }

        @NotNull
        @Override
        public TSMetaEnum getOwner() {
            return myOwner;
        }
    }
}

File: src/com/intellij/idea/plugin/hybris/type/system/meta/model/impl/TSMetaEnumImpl.kt
/*
 * This file is part of "SAP Commerce Developers Toolset" plugin for Intellij IDEA.
 * Copyright (C) 2019 EPAM Systems <hybrisideaplugin@epam.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as 
 * published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.intellij.idea.plugin.hybris.type.system.meta.model.impl

import com.intellij.idea.plugin.hybris.type.system.meta.impl.CaseInsensitive
import com.intellij.idea.plugin.hybris.type.system.meta.model.TSGlobalMetaEnum
import com.intellij.idea.plugin.hybris.type.system.meta.model.TSMetaEnum
import com.intellij.idea.plugin.hybris.type.system.meta.model.TSMetaEnum.TSMetaEnumValue
import com.intellij.idea.plugin.hybris.type.system.meta.model.TSMetaSelfMerge
import com.intellij.idea.plugin.hybris.type.system.model.EnumType
import com.intellij.idea.plugin.hybris.type.system.model.EnumValue
import com.intellij.openapi.module.Module
import com.intellij.util.xml.DomAnchor
import com.intellij.util.xml.DomService

internal class TSMetaEnumImpl(
    dom: EnumType,
    override val module: Module,
    override val name: String?,
    override val isCustom: Boolean,
    override val values: Map<String, TSMetaEnumValue>
) : TSMetaEnum {

    override val domAnchor: DomAnchor<EnumType> = DomService.getInstance().createAnchor(dom)
    override val isAutoCreate = java.lang.Boolean.TRUE == dom.autoCreate.value
    override val isGenerate = java.lang.Boolean.TRUE == dom.generate.value
    override val isDynamic = java.lang.Boolean.TRUE == dom.dynamic.value
    override val description = dom.description.stringValue
    override val jaloClass = dom.jaloClass.stringValue

    override fun toString() = "Enum(module=$module, name=$name, isDynamic=$isDynamic, isCustom=$isCustom)"

    internal class TSMetaEnumValueImpl(
        dom: EnumValue,
        override val module: Module,
        override val isCustom: Boolean,
        override val name: String
    ) : TSMetaEnumValue {

        override val domAnchor: DomAnchor<EnumValue> = DomService.getInstance().createAnchor(dom)
        override val description = dom.description.stringValue

        override fun toString() = "EnumValue(module=$module, name=$name, isCustom=$isCustom)"
    }

}

internal class TSGlobalMetaEnumImpl(localMeta: TSMetaEnum)
    : TSMetaSelfMerge<EnumType, TSMetaEnum>(localMeta), TSGlobalMetaEnum {

    override val values = CaseInsensitive.CaseInsensitiveConcurrentHashMap<String, TSMetaEnumValue>()
    override val domAnchor = localMeta.domAnchor
    override val module = localMeta.module
    override var isAutoCreate = localMeta.isAutoCreate
    override var isGenerate = localMeta.isGenerate
    override var isDynamic = localMeta.isDynamic
    override var description = localMeta.description
    override var jaloClass = localMeta.jaloClass

    override fun mergeInternally(localMeta: TSMetaEnum) {
        jaloClass?:let { jaloClass = localMeta.jaloClass }
        description?:let { description = localMeta.description }

        if (localMeta.isDynamic) isDynamic = localMeta.isDynamic
        if (localMeta.isAutoCreate) isAutoCreate = localMeta.isAutoCreate
        if (localMeta.isGenerate) isGenerate = localMeta.isGenerate

        localMeta.values.values
            .filterNot { values.contains(it.name) }
            .forEach { values[it.name] = it }
    }

    override fun toString() = "Enum(module=$module, name=$name, isDynamic=$isDynamic, isCustom=$isCustom)"
}